[CI] JdbcSqlSpecIT fails: "Expected more data..."

https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+multijob-unix-compatibility/os=oraclelinux/2389/console

REPRODUCE WITH lines both reproduce a failure for me:

```
./gradlew :x-pack:qa:sql:security:ssl:integTestRunner \
  -Dtests.seed=1118CB444CAB2365 \
  -Dtests.class=org.elasticsearch.xpack.qa.sql.security.JdbcSqlSpecIT \
  -Dtests.method="test {agg.testAggMultiWithHaving}" \
  -Dtests.security.manager=true \
  -Dtests.locale=ko-KR \
  -Dtests.timezone=Asia/Jerusalem

./gradlew :x-pack:qa:sql:security:ssl:integTestRunner \
  -Dtests.seed=1118CB444CAB2365 \
  -Dtests.class=org.elasticsearch.xpack.qa.sql.security.JdbcSqlSpecIT \
  -Dtests.method="test {agg.testAggMultiWithHavingOnCount}" \
  -Dtests.security.manager=true \
  -Dtests.locale=ko-KR \
  -Dtests.timezone=Asia/Jerusalem
```

```
   > Throwable #1: java.lang.AssertionError: Expected more data but no more entries found after [0]
   >    at __randomizedtesting.SeedInfo.seed([1118CB444CAB2365:994CF49EE2574E9D]:0)
   >    at org.elasticsearch.xpack.qa.sql.security.JdbcSqlSpecIT.agg.testAggMultiWithHaving(agg.sql-spec:279)
   >    at org.elasticsearch.xpack.qa.sql.jdbc.JdbcAssert.doAssertResultSetData(JdbcAssert.java:114)
   >    at org.elasticsearch.xpack.qa.sql.jdbc.JdbcAssert.assertResultSetData(JdbcAssert.java:103)
   >    at org.elasticsearch.xpack.qa.sql.jdbc.JdbcAssert.assertResultSets(JdbcAssert.java:37)
   >    at org.elasticsearch.xpack.qa.sql.jdbc.SpecBaseIntegrationTestCase.assertResults(SpecBaseIntegrationTestCase.java:103)
   >    at org.elasticsearch.xpack.qa.sql.jdbc.SqlSpecTestCase.doTest(SqlSpecTestCase.java:67)
   >    at org.elasticsearch.xpack.qa.sql.jdbc.SpecBaseIntegrationTestCase.test(SpecBaseIntegrationTestCase.java:76)
   >    at java.lang.Thread.run(Thread.java:748)
```

--- a/x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/execution/search/CompositeAggregationCursor.java
+++ b/x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/execution/search/CompositeAggregationCursor.java
@@ -113,12 +113,36 @@ public class CompositeAggregationCursor implements Cursor {
 
         SearchRequest search = Querier.prepareRequest(client, query, cfg.pageTimeout(), indices);
 
-        client.search(search, ActionListener.wrap(r -> {
-            updateCompositeAfterKey(r, query);
-            CompositeAggsRowSet rowSet = new CompositeAggsRowSet(extractors, r, limit,
-                    serializeQuery(query), indices);
-            listener.onResponse(rowSet);
-        }, listener::onFailure));
+        client.search(search, new ActionListener<SearchResponse>() {
+            @Override
+            public void onResponse(SearchResponse r) {
+                try {
+                    // retry
+                    if (shouldRetryDueToEmptyPage(r)) {
+                        CompositeAggregationCursor.updateCompositeAfterKey(r, search.source());
+                        client.search(search, this);
+                        return;
+                    }
+
+                    updateCompositeAfterKey(r, query);
+                    CompositeAggsRowSet rowSet = new CompositeAggsRowSet(extractors, r, limit, serializeQuery(query), indices);
+                    listener.onResponse(rowSet);
+                } catch (Exception ex) {
+                    listener.onFailure(ex);
+                }
+            }
+
+            @Override
+            public void onFailure(Exception ex) {
+                listener.onFailure(ex);
+            }
+        });
+    }
+
+    static boolean shouldRetryDueToEmptyPage(SearchResponse response) {
+        CompositeAggregation composite = getComposite(response);
+        // if there are no buckets but a next page, go fetch it instead of sending an empty response to the client
+        return composite != null && composite.getBuckets().isEmpty() && composite.afterKey() != null && !composite.afterKey().isEmpty();
     }
 
     static CompositeAggregation getComposite(SearchResponse response) {
@@ -134,7 +158,7 @@ public class CompositeAggregationCursor implements Cursor {
         throw new SqlIllegalArgumentException("Unrecognized root group found; {}", agg.getClass());
     }
 
-    static void updateCompositeAfterKey(SearchResponse r, SearchSourceBuilder next) {
+    static boolean updateCompositeAfterKey(SearchResponse r, SearchSourceBuilder next) {
         CompositeAggregation composite = getComposite(r);
 
         if (composite == null) {
@@ -152,7 +176,10 @@ public class CompositeAggregationCursor implements Cursor {
             } else {
                 throw new SqlIllegalArgumentException("Invalid client request; expected a group-by but instead got {}", aggBuilder);
             }
+
+            return true;
         }
+        return false;
     }
 
     /**

--- a/x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/execution/search/CompositeAggregationCursor.java
+++ b/x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/execution/search/CompositeAggregationCursor.java
@@ -158,7 +158,7 @@ public class CompositeAggregationCursor implements Cursor {
         throw new SqlIllegalArgumentException("Unrecognized root group found; {}", agg.getClass());
     }
 
-    static boolean updateCompositeAfterKey(SearchResponse r, SearchSourceBuilder next) {
+    static void updateCompositeAfterKey(SearchResponse r, SearchSourceBuilder next) {
         CompositeAggregation composite = getComposite(r);
 
         if (composite == null) {
@@ -176,10 +176,7 @@ public class CompositeAggregationCursor implements Cursor {
             } else {
                 throw new SqlIllegalArgumentException("Invalid client request; expected a group-by but instead got {}", aggBuilder);
             }
-
-            return true;
         }
-        return false;
     }
 
     /**

--- a/x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/execution/search/Querier.java
+++ b/x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/execution/search/Querier.java
@@ -206,8 +206,15 @@ public class Querier {
         protected void handleResponse(SearchResponse response, ActionListener<SchemaRowSet> listener) {
             // there are some results
             if (response.getAggregations().asList().size() > 0) {
-                CompositeAggregationCursor.updateCompositeAfterKey(response, request.source());
 
+                // retry
+                if (CompositeAggregationCursor.shouldRetryDueToEmptyPage(response)) {
+                    CompositeAggregationCursor.updateCompositeAfterKey(response, request.source());
+                    client.search(request, this);
+                    return;
+                }
+
+                CompositeAggregationCursor.updateCompositeAfterKey(response, request.source());
                 byte[] nextSearch = null;
                 try {
                     nextSearch = CompositeAggregationCursor.serializeQuery(request.source());

--- a/x-pack/qa/sql/security/src/test/java/org/elasticsearch/xpack/qa/sql/security/JdbcSqlSpecIT.java
+++ b/x-pack/qa/sql/security/src/test/java/org/elasticsearch/xpack/qa/sql/security/JdbcSqlSpecIT.java
@@ -5,13 +5,11 @@
  */
 package org.elasticsearch.xpack.qa.sql.security;
 
-import org.apache.lucene.util.LuceneTestCase.AwaitsFix;
 import org.elasticsearch.common.settings.Settings;
 import org.elasticsearch.xpack.qa.sql.jdbc.SqlSpecTestCase;
 
 import java.util.Properties;
 
-@AwaitsFix(bugUrl = "https://github.com/elastic/elasticsearch/issues/30292")
 public class JdbcSqlSpecIT extends SqlSpecTestCase {
     public JdbcSqlSpecIT(String fileName, String groupName, String testName, Integer lineNumber, String query) {
         super(fileName, groupName, testName, lineNumber, query);

